templates/coredns/coredns.py

```diff
@@ -0,0 +1,119 @@
+        self.recurring_action('_monitor', 30)  # every 30 seconds
+
+    def _monitor(self):
+        self.logger.info('Monitor coredns %s' % self.name)
+        self.state.check('actions', 'start', 'ok')
+
+        if not self._coredns_sal.is_running():
+            self.state.delete('status', 'running')
+            self._coredns_sal.deploy()
+            self._coredns_sal.start()
+            if self._coredns_sal.is_running():
+                self.state.set('status', 'running', 'ok')
+        else:
+            self.state.set('status', 'running', 'ok')
+    @property
+    def _node_sal(self):
+        """
+        connection to the node
+        """
+        return j.clients.zos.get(NODE_CLIENT)
+
+    @property
+    def _node(self):
+        if not self._node_:
+            self._node_ = self.api.services.get(template_account='threefoldtech', template_name='node')
+        return self._node_
+
+    @property
+    def _coredns_sal(self):
+        kwargs = {
+            'name': self.name,
+            'node': self._node_sal,
+            'recursive_resolvers':self.data['upsteram'],
+            'zt_identity': self.data['ztIdentity'],
+            'nics': self.data['nics'],
+            'etcd_endpoint': self._etc_url
+        }
+        return j.sal_zos.coredns.get(**kwargs)
+    @property
+    def _etcd(self):
+        return self.api.services.get(template_uid=ETCD_TEMPLATE_UID, name=self.data['etcdServerName'])
+    @property
+    def _etc_url(self):
+        result = self._etcd.schedule_action('connection_info').wait(die=True).result
+        return result['client_url']
+        
+    def node_port(self):
+        return self.data['nodePort']
+
+    def install(self):
+        self.logger.info('Installing CoreDns %s' % self.name)
+
+        coredns_sal = self._coredns_sal
+
+        coredns_sal.deploy()
+        self.data['ztIdentity'] = coredns_sal.zt_identity
+
+        self.data['nodePort'] = coredns_sal.node_port
+        self.logger.info('Install CoreDns %s is Done' % self.name)
+        self.state.set('actions', 'install', 'ok')
+
+    def uninstall(self):
+        self.logger.info('Uninstalling CoreDns %s' % self.name)
+        self._coredns_sal.destroy()
+        self.state.delete('actions', 'install')
+        self.state.delete('status', 'running')
+
+    def start(self):
+        """
+        start coredns server
+        """
+        self.state.check('actions', 'install', 'ok')
+        self.logger.info('Starting CoreDns  %s' % self.name)
+        self._coredns_sal.deploy()
+        self._coredns_sal.start()
+        self.state.set('actions', 'start', 'ok')
+        self.state.set('status', 'running', 'ok')
+
+    def stop(self):
+        self.state.check('actions', 'install', 'ok')
+        self.logger.info('Stopping CoreDns %s' % self.name)
+        self._coredns_sal.stop()
+        self.state.delete('actions', 'start')
+        self.state.delete('status', 'running')
+
+    def register_domain(self, domain, ip, ttl=300):
```

Comment:
I'm wondering if domain shoudn't be a service by iself instead of some methods here. @serboctor what do you think ?

Reply:
Hmm I dont think it is really worth it. But may be you're seeing something I am not. Why do you think it should be a service?

Reply:
Cause the coredns will most probably be a public service that can be used by different user.
But then, the domain service can be owned by different clients and they will own their domain when they created it.
If we don't have that separation, only the owner of the coredns service will be able to add domain to it

Reply:
Ok so I've got more information about all this. It is another component that will connect to etcd and handle the management of the domain and virtual host. So there is no need to have such method on the services themself